modin/core/execution/ray/common/deferred_execution.py

```diff
@@ -0,0 +1,823 @@
+        to the remote method as is.
+    flat_kwargs : bool, optional
+        The same as `flat_args` but for the `kwargs` values.
+    """
+
+    def __init__(
+        self,
+        data: Union[
+            ObjectRefType,
+            "DeferredExecution",
+            List[Union[ObjectRefType, "DeferredExecution"]],
+        ],
+        func: Union[Callable, ObjectRefType],
+        args: Union[List[Any], Tuple[Any]],
+        kwargs: Dict[str, Any],
+        num_returns=1,
+        flat_args: Optional[bool] = None,
+        flat_kwargs: Optional[bool] = None,
+    ):
+        ref = DeferredExecution._ref
+        if flat_args is None:
+            if has_list_or_de(args):
+                flat_args = False
+                for a in args:
+                    ref(a)
+            else:
+                flat_args = True
+        if flat_kwargs is None:
+            if has_list_or_de(kwargs.values()):
+                flat_kwargs = False
+                for a in kwargs.values():
+                    ref(a)
+            else:
+                flat_kwargs = True
+        ref(data)
+        self.data = data
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+        self.flat_args = flat_args
+        self.flat_kwargs = flat_kwargs
+        self.num_returns = num_returns
+        self.ref_counter = 0
+
+    @staticmethod
+    def _ref(obj):
+        """
+        Increment the `ref_counter` if `obj` is a `DeferredExecution`.
+
+        Parameters
+        ----------
+        obj : Any
+        """
+        if isinstance(obj, DeferredExecution):
+            obj.ref_count(1)
+        elif isinstance(obj, ListOrTuple):
+            ref = DeferredExecution._ref
+            for o in obj:
+                ref(o)
+
+    def exec(
+        self,
+    ) -> Tuple[ObjectRefOrListType, Union["MetaList", List], Union[int, List[int]]]:
+        """
+        Execute this task, if required.
+
+        Returns
+        -------
+        tuple
+            The execution result, MetaList, containing the length, width and
+            the worker's ip address (the last value in the list) and the values
+            offset in the list. I.e. length = meta_list[offset],
+            width = meta_list[offset + 1], ip = meta_list[-1].
+        """
+        if self.has_result:
+            return self.data, self.meta, self.meta_off
+
+        data = self.data
+        if self.num_returns == 1 and not isinstance(data, DeferredExecution):
+            if self.flat_kwargs and self.flat_kwargs:
+                result, length, width, ip = remote_exec_func.remote(
+                    self.func, data, *self.args, **self.kwargs
+                )
+                meta = MetaList([length, width, ip])
+                self._set_result(result, meta, 0)
+                return result, meta, 0
+
+        consumers, output = self._deconstruct()
+        num_returns = sum(c.num_returns for c in consumers) + 1
```

Comment:
```suggestion
        offset = 1
        num_returns = sum(c.num_returns for c in consumers) + offset
```
IIUC. Just for better reading